naming/program/src/state.rs

```diff
@@ -0,0 +1,47 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use solana_program::{msg, program_error::ProgramError, pubkey::Pubkey};
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone, Copy)]
+pub enum NameStatus {
+    Deleted,
+    Active,
+}
+
+pub type ValueType = Pubkey;
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone, Copy)]
+pub enum IssuerKind {
+    Metadata,       // Metadata authority has sole control
+    Issuer(Pubkey), // Issuing authority has sole control
+    Both(Pubkey),   // Issuing and metadata authority must sign
+    Either(Pubkey), // Issuing or metadata authority must sign
+    Open,           // First come, first served. Terminal state
+    Frozen,         // No further issuance allowed.  Terminal state
+}
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone)]
+pub enum OwnerKind {
+    Account(Pubkey),
+    Name(String), // Usefull for easy ownership transfer of a whole sub-tree of domain names
+    Frozen,       // Metadata is locked, terminal state
+}
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone)]
+pub struct NameRegistryState {
+    pub status: NameStatus,
+    // This is the complete name, including all parents
+    pub name: String,
```

Comment:
1. Including the parents in the name seems redundant IDK
2. I don't think we need to include the name at all :).  The name is already known because it was included in the address derivation hash.  For name discovery, I was envisioning a name record would have its own account (one value per account as per my comment below) where the "value data" is the name itself.